plugin/hooks.go

```diff
@@ -155,4 +156,10 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// NotificationWillBeSent is invoked before a notification is sent.
+	// To modify the notified user map, return the replacement map,
+	//
+	// Minimum server version: 5.20
+	NotificationWillBeSent(c *Context, post *model.Post, mentions *model.ExplicitMentions) *model.ExplicitMentions
```

Comment:
I'm wondering if we also should add a `NotificationHasBeSent` while touching this code piece. I should not be a lot of effort. @rfoyard Thoughts?

Reply:
I'm not sure I can find a use case for this, but it should be possible.  
The thing is there are different type of notification that can be sent, a user can receive an email and a push notification and an ephemeral post (at the same time !).  
They all are using a different code path and I'm not sure how to deduplicate them.  
Do you see any use case for such a callback ?

Reply:
There might be the use case to mirror notification to an external application .

But this is not a blocking request. If it's hard to implement, just leave it out.

Reply:
IMO, 4/5 make a separate ticket for that, since it would probably involve asking "who needs that hook, and why" and "would that hook run on every type of notification, or just x?" etc.